ADA/Guestbook-at-request

qqmyers opened the pull request:
**What this PR does / why we need it**: This is a restart of #6902 which adds optional support for guestbooks to appear when files access is requested rather than after access has been granted and a download is started. This is key functionality ADA needs to be able to adopt Dataverse 5.x+.

The PR currently compiles but is not yet known to be working. (Making the PR now to make the progress more visible and to be able to monitor the automated test runs.)

**Which issue(s) this PR closes**:

Closes #2043 
Closes #9701

**Special notes for your reviewer**:

**Suggestions on how to test this**:

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

**Is there a release notes update needed for this change?**:

**Additional documentation**:

mdmADA commented:
Hi Jim.

Thanks for you work on this. I was able to test the AWS instance specified in your email of June 24, 2023.

**Question:** Will the location in the access request worklfow be configurable? Ideally, it would be possible to specify for each dataset where the guestbook should be presented to the end user:
- at file download (default, the current behaviour)
- at guestbook request (for those, like ADA, who want the guestbook at this point)

There is a use case for ADA where 97% of datasets require the guestbook at request access and the other 3% require the guestbook at file download, so being able to configure this would be ideal.

**Clarification:**  The access request status (CREATED, etc) that you referred to in your email of June 24 is not needed. I know it is in my code but I had indicated in one of the community calls that this is no longer required, at least at the current time. It makes things more complicated than they need to be so please ignore that part of the code.

**TESTING RESULTS**
I created "Test Dataset 1" with one user, created 3 test data files and restricted them, and added a guestbook to the dataset.
I requested access with another user.

BUGS:
1. As the requester, when I click the arrow dropdown menu to the right of one of the restricted files and click 'Request Access', nothing happens. The guestbook does not pop up. (Checking checkboxes to the left of the files and clicking the multi-select 'Request Access' does pop up the guestbook).

2. Upon submission of the guestbook details by clicking 'Request Access' in the popup, everything works correctly but there are 3 things missing:
- a new Notification in the requester's Notifications that says "You have requested access to 'Test Dataset 1' ... "
- an email to the requester saying "You have requested access to 'Test Dataset 1' ... "
- an email to the dataset owner saying "File access requested for 'Test Dataset 1' ..."

3. Once the dataset owner grants access to any or all of the files requested, there is still a row in the 'Requests' table of the dataset's permissions table (albeit with 0 Files). I assume this is due to the record in the fileaccessrequests table being assigned status 'CREATED' and is not deleted from the table. As the access request status is no longer required, the original behaviour of deleting the row from fileaccessrequests should be reinstated.

<img width="1369" alt="Screenshot - entries in both Requests and Users:Groups" src="https://github.com/IQSS/dataverse/assets/23644215/300b20e0-bc41-470a-b3bf-19ffe2548433">

4. When the requester goes back to the dataset after being granted access to the datafiles, the message 'Access requested' still shows in the files table and in the files' individual dropdown menus (and in the single file page's dropdown menu). (Probably due to the rows still being in fileaccessrequests).

5. Revoking access from the data files should reset them to be 'requestable' to the revokee but they are showing as 'Access requested' and can't be requested. (Probably due to the rows still being in fileaccessrequests).

<img width="1324" alt="Screenshot - revoked should be requestable again" src="https://github.com/IQSS/dataverse/assets/23644215/da73312f-bc25-4db3-90de-87c184ede1c6">

Thank you.

qqmyers commented:
@mdmADA - Thanks! I've worked through your list of bugs, added in the ability to configure this globally and per collection/dataset, made it work for requests from the file page, and added some documentation. A few more notes below, but unless you find further issues or have other change requests I think this can go to review/QA. (Right now, the expectation is that this would go into a ~6.1 release as 5.14/6.0 are getting frozen and this should have a good amount of testing for regression.)

- The notification/email to the dataset contacts should be working. There is currently no notification/email to the person making the request (nominally they know since they just took the action). There is discussion about this being a useful future addition - see #9701.
- I simplified the states to only CREATED, GRANTED, REJECTED - I thought that was useful for accountability - as a record of whether a given request with its associated guestbook entry was initially granted/rejected regardless of subsequent changes to the permissions. Things *should* work correctly such that you can re-request after a rejection, access requested only shows when the request is still in the CREATED state, only CREATED requests show in the manage file permissions page, etc.
- The global setting determines the overall default and enables the ability to change the setting for collections/datasets. (If the global setting isn't set, the behavior remains as guestbook at download and can't be changed. Setting it false provides a global default of guestbook at download but would allow that to be changed  for collections/datasets.) As with other settings, the guestbook at request/at download choice is inherited down through sub-collections and can be changed in the Dataverse/Collection page Edit/General Settings panel. Also as with other settings, the ability to set this for a given dataset is only allowed via API and only by a superuser. The reason it is API only is just because there isn't an existing UI for dataset options as there is for collections (so we're avoiding that work). The reason it is superuser is mostly for consistency with other, more sensitive settings. I think this could be dropped to allow anyone who can edit the dataset to change the setting, so if that is part of your use case, let me know.

mdmADA commented:
Hi Jim.

These are the details of testing with guestbook at request access and at file download. There are a few (BUG)s as indicated in the description. Hopefully it's easy to understand:

**A. TESTING GUESTBOOK AT REQUEST ACCESS**
Use: Test GB at RA Dataverse

Created guestbook and assigned it to dataset
Created 3 files, all restricted

**Test A1:**
Dataverse setting: Guestbook at Request Access

	1. dataset.xhmtl:
	The terms of access + guestbook details pop up when the multi-select and individual file dropdown -> Request Access
		-> correct
		
	2. file.xhtml page
	The terms of access + guestbook details pop up when file dropdown -> Request Access clicked
		-> correct


**Test A2:**
change: Dataverse setting to Guestbook at File Download

1. dataset.xhtml page:
	
	a. Select one or more multi-select file checkboxes
	Multi-select request access
	-> only the terms of access show (correct) 
	
	b. Select the file dropdown menu to the right of a single file
	click request Access option
	-> only the terms of access show (correct) 
	

2. Click the file link to go to the file.xhtml page
	
	click the "Access File" file dropdown menu
	Click the request access option
	-> the terms of access + the guestbook details show **(BUG)**
	
	

**Test A3:**
Now: unrestrict 1 of the files (Guestbook at Request Access)

1. dataset.xhtml & file.xhmtl

- go through steps to download the unresticted file
- only access terms are shown (no gb) -> accept -> downloads

This is how ADA's dataverse works - the guestbook applies to only restricted files but I _am not sure if this is what is expected from the Dataverse team's pov_


	
	
**B. TESTING GUESTBOOK AT FILE DOWNLOAD**
Use: Test GB at File Download Dataverse

Created guestbook and assigned it to dataset
Created 3 test files, 1 open, 2 restricted
Login as requester

**Test B1:**
Dataverse setting: Guestbook Entry at File Download

1. Download the unrestricted file
- guestbook pops up 
- I don't know if this should be the behaviour
	- should the gb be only for when restricted files are downloaded after access granted??
	
	NOTE: testing on demo.dataverse (v5.13)
		- dataset with assigned guestbook
		- 1 unrestricted, 2 restricted files
		- guestbook pops up for the open file so the default behaviour is to enter guestbook details when unrestricted file is downloaded
		- just to note that when the guestbook is at request access, the guestbook does not pop up when unrestricted files are downloaded.
	
**Test B2:**
Dataverse setting: Guestbook Entry at File Download
		
1. dataset.xhtml page
- clicking request access with Multi-select Request Access or single file request Access
	- only shows the terms of access
	- does not show the guestbook (correct)
	- request submitted successfully
	
	- Possible BUG: when the user clicks request access, the request is sent
		login as the dataset owner and go to the dataverse -> guestbooks -> view the guestbook responses
				- the guestbook details are still submitted but with empty values
				- this is either a bug or ignore those empty rows


**Test B3:**
Dataverse setting: Guestbook Entry at File Download

2. click on restricted file link in dataset page -> file.xhtml page
- click Request Access in menu dropdown
- displays guestbook (same scenario as in A2.2 above)  (BUG)
	

**Test B4:**
Dataverse setting: Guestbook Entry at File Download
	
3. Login as dataset owner and grant access to the requester.
	Requester can now download.
	The guestbook pops up when the user downloads the file.
	
	(The response also appears in the dataverse -> guestbooks -> view guestbook responses which is correct behaviour)	
	
	
Note: No notifications are created in the requester's ui in all of the tests above - there should be **(BUG)**.

mdmADA commented:
Regarding the API ENDPOINT to set the guestbook point for a specific dataset:

For the API call to set the guestbook popup point for a specific dataset, has that API endpoint been created?
If yes, what's the API endpoint and its parameters?

Can you assign my "test" user to the correct permissions so I can see if I can set the gb popup point that way?

For users who have permissions to edit the dataset, will there be any way for them to know if the gb point has been set to "file download" or "request access", or will it require a database query to determine this? (I am thinking of data archivists who take over datasets from someone who has departed and who would need to know if it's been set properly or not.)

mdmADA commented:
Regarding whether to send an email to the requester:	
	
All those years back I added the functionality to ADA's instance that when a dataset's files are requested that a much more detailed email is sent to the dataset contact (ADA in 99% of cases). I just realised yesterday that this detailed email won't be in your code.

To explain, this detailed email is used to create a ticket for ADA's Access team, and all of the details of the request are in the ticket including the requester's details and the details of the guestbook response they entered in the request. 

I envision that this same email would also be sent to the requester.

This much more detailed email being sent to the requester has a few purposes from ADA's point of view:
1. consistency in that emails are sent to dataset contact so also send one to the person who requested it
2. as a confirmation to the requester that the request indeed went through (users might not notice any new Notifications in the Dataverse ui if request notifications are )
3. as a record for the user as to the details they submitted - as far as I know, there is no way for the requester to ever see these details in Dataverse again
	- they can refer to the email and details later
	- they can reuse the details to copy and paste into any new requests rather than having to retype all the info again


I had asked in 2019 (issue #6222) if it would be possible to allow templates to be specified for emails to be sent to different users. Would it be difficult to implement this? Or implement the longer email to be sent?

qqmyers commented:
Re: the api: details are at https://dataverse-guide--9599.org.readthedocs.build/en/9599/api/native-api.html#configure-when-a-dataset-guestbook-appears-if-enabled. I've given you superuser access so you can set the option per dataset. FWIW: The GET call doesn't require special permissions to use, so anyone can check what the setting is that way. (still looking at your other comments).

stevenmce commented:
Hi all,

You might want to take a look at the recent #9701 feature request - there is some overlap here. ADA sends a confirmation email on request to the user. As @mdmADA notes, we would also want a notification to the data owner/decision maker with the details of the request.

Cheers,
Steve

mdmADA commented:
@stevenmce - Jim references issue #9701 in his second comment from July 13 or 14.
The primary issue for ADA (as you indicate in your comment above) is for the detailed email to be sent to ADA as the dataset contact/decision maker.

qqmyers commented:
I think all the bugs have been addressed and I've added a pretty-printed version of the guestbook response in the mail to the dataset owner. 

re: Test A3/B1 - I think not showing the guestbook when downloading an unrestricted file is OK, although, as you say,  different from when gb at download is enabled. (Note that the terms actually only show if you are a)not using CC0, or b) have Terms of Access set, otherwise you just get the download directly. (Since Terms of Access only apply to restricted files, I'm not sure they should show for an unrestricted file. This is how it works now though.))

The difference in behavior between at request and at download seems OK to me as long as it is documented. The one Q I have is whether you thing the gb should show at download in the at request case when file access requests are disabled. I think that is how you had it in your original branch. Fine if needed but it might be more consistent with the unrestricted case if at request only shows the gb at request and never at download, regardless of whether the file is unrestricted or you've been given access. Let me know if that should change.

Re: Note: No notifications are created in the requester's ui in all of the tests above - there should be (BUG). 
I'm not sure what you mean - is this a change from v5.13 behavior or just part of the #9701-related request to add a new notification/email?

Re: adding gb info to the email to admins - if we need a different format, let me know. pretty JSON was easy to create and is both machine and human readable, but it could be CSV, etc. instead. I also have not included everything about the gb request, just mainly the questions and answers. If you need other attributes included, let me know.

mdmADA commented:
I am sorry to do this to you right before the weekend. 

I tested the code as it is now on the AWS instance.

**1. GB at Request Access Dataverse - Guestbook Entry at Request Access**

- For both datasets, there is a guestbook applied to each
- For both datasets:
	- the guestbook only appears when multiple files selected and the Mult-Select "Request Access" button clicked
	- the guestbook isn't appearing when the single file dropdown menu 'Request Access' selected **(BUG)**
	- the guestbook isn't appearing in the single file.xhtml page's Access File -> Request Access clicked **(BUG)**


**2. GB at Download Dataverse - Guestbook at Download**

The guestbook isn't popping up when 'Request Access' clicked:
- no guestbook on dataset.xhtml for multiselect 'Request Access" and for single file 'Request Access' (correct)
- no guestbook on file.xhtml Access File -> Request Access (correct)

**Problems:**
**2A. Click single file link**
Takes you to file.xhtml page
Request Access in the single page
	- Access Requested shows in the Access Dataset dropdown menu
Use the browser back navigation
"Access Requested" does not show for the file in the files table
"Access Requested" does not show in the dropdown menu
User can request again **(BUG)**
	- not sure what happens in the backend - write error to fileaccessrequests table?
	
	
**2B. Data owner grants access to the requested files**

User to whom access was granted:
From the dataset page:

Guestbook pops up when multi-select download used (correct)
Guestbook pops up for each of the restricted file's single dropdown menu download (correct)

Go to single file.xhtml page - guestbook does not pop up on download **(BUG)**

mdmADA commented:
Regarding the creation of the notification for the user who requested access, it may be the behaviour that I implemented for ADA, but if it's not in iqss Dataverse code, I suggest that a notification is created for the user who requested the data: "You have requested access to <dataset name as link>". Or similar. So the user can see in their Notifications what datasets they have requested access to, rather than having to navigate to a specific dataset to see the "Access requested" messages.

This notification would also mirror the "File access requested for <dataset> by <user>" notification that is created for the dataset contact/owner on the dataset admin side of things.

mdmADA commented:
Regarding the discussion about what should happen/the difference in behaviour when there are unrestricted files with gb at download and at request access, maybe that's a discussion for the IQSS team as what should happen? Or with the wider dataverse community?

When I implemented the code for ADA way back in 2017, I was of the understanding that 100% of ADA's datasets would be guestbook at request access so the no guestbook at download is what happens for the unrestricted files (usually documentation like codebooks, etc.) for a dataset with a guestbook. I don't know if that's what should happen though, from IQSS's perspective.

qqmyers commented:
@mdmADA - re the recent bug list: Sorry - I think this was just out-of-date code. I think I saw one of the issues you mentioned this morning and then updated the code and it and the others were gone. I've since merged with develop again and I think things are still working. If not, give me details about which user and datasets are involved. (I'm mostly using the GBAR dataset and using the API to flip the gb at request option for it.)

As for email, I'll go ahead and add one for the requestor. I think it would be better someday to have a page displaying requests and their statuses, but that's definitely more work (and work that would have to be redone for the SPA), so email it is.

re: unrestricted files - since this PR is leaving the gb at download functionality as is, I think it is fine to define that when using gb at request there is no gb at download for unrestricted files. (I would think that a GB that is rigorous to allow making decisions about whether to grant access to restricted files would be overkill for getting unrestricted files like readme/documentation.) If IQSS feels that should change when they review, I can flip it.

mdmADA commented:
In testing again (the GB at RA and the GB at FD):

For gb at ra: 
Page.xhtml -> gb appears with the multi-select and single file request access (correct)
File.xhtml -> gb appears with Access File -> request access (correct)
After being granted access, the gb does not appear at download (correct).

For gb at fd:
Page.xhtml -> gb does not appear at request access for both multi-select and single file (correct)
File.xhtml -> gb does not appear with Access File -> request access (correct)
(Note that an access request with blank guestbook response values is still submitted)
After being granted access, the gb does appear at download (correct).

**************************

**However,** there seems to be an issue with the dataset.xhtml page for both (gb at ra) and (gb at fd) with the **"Access Dataset" button.**

GB at RA:
1. After being granted access to Test Dataset 1 files, click "Access Dataset" -> Original => gb pops up
2. Go down the page to the files table and click one of the single file download or multi-select downloads -> no gb (correct)
3. Go back up the page and click "Access Dataset" - Original and now no gb
Reload the page and you are back at 1. again. 

So the "Access Dataset" button seems to be affected by the other buttons in the page.

Similar behaviour occurs for the GB at FD and the "Access Dataset" button and the other buttons in dataset.xhmtl.
1. Click the "Access Dataset" -> Original and no gb appears
2. Click either the multi-select Download or single file download -> Original -> gb appears at download
3. Go back and click the Access Dataset button -> Original -> gb appears at download
Reload page and you are back at 1 (if "Access Dataset" clicked first).

mdmADA commented:
I missed responding to your previous question: "The one Q I have is whether you thing the gb should show at download in the at request case when file access requests are disabled. I think that is how you had it in your original branch. Fine if needed but it might be more consistent with the unrestricted case if at request only shows the gb at request and never at download, regardless of whether the file is unrestricted or you've been given access. Let me know if that should change."

I think that if the setting is gb at ra, then the gb should only show at request access and never at download, even for unrestricted files. This includes when File Access request is disabled. (This is how it currently works for ADA as you can see here: https://dataverse.ada.edu.au/dataset.xhtml?persistentId=doi:10.4225/87/QFUIBM)

mdmADA commented:
Finally - I think! - I am copying and pasting the format of the email that ADA's Dataverse instance has been modified to send to ADA's email, which results in the creation of a ticket for ADA's access team.  Perhaps a json pretty print would be sufficient. It's only informative though if both the questions and responses are contained.  

An email to the dataset contact/owner that contains the above details is a requirement for ADA as that email is what creates tickets in ADA's ticketing system for ADA's access management team. 

An email to the requester with the guestbook questions and responses would also be beneficial for reasons given previously (and in lieu of the requests page functionality that seems to far down the roadmap for the Dataverse SPA).


Email format:

Subject: ADA Dataverse: "<firstname lastname (@useridentifier)>" requested access to dataset "<dataset title>"

Hello,

File access requested for dataset: <dataset title>. Manage permissions at <link to dataset's permissions-manage-file.xhtml>

Thank you,
The Australian Data Archive
on behalf of: <list of dataset contacts/owners>

*******************************************    

(followed by appending of the user's guestbook response details. Example:)

ACCESS REQUEST GUESTBOOK DETAILS:

guestbookresponse_id: dddddd
authenticateduser_id: dddddd
@<useridentifer> has verified their email address: yes

Name: <name as entered>
Email: <email as entered>
Institution or Organisation: <Institution as entered>
Position: <Position as entered>


Custom Questions:

What is your primary intended use of this data?: <response as entered>

If other uses, please indicate what other intended use you have for this data?: ----- (optional question, nothing entered)

Please indicate the name and contact email of your supervisor: ----- (optional question, nothing entered)

Please provide a brief abstract outlining the intended use of this data for your project: <response as entered>

Please detail any sources of funding supporting this research?: <response as entered>

What are the intended publications and outputs that may feature this data?: <response as entered>

qqmyers commented:
- I've fixed the dataset-level download buttons.
- I could not reproduce having an empty guestbook entry for access requests when gb at download is on.
- I've added an email to the requestor, including the gb entry if gb-at-request. I've updated the format to be textual rather than JSON and am now sending the same format in the emails to the managers.
- Also restored the success/fail messages at the top of the page
- Added i18n for the GB response in the emails.
- Removed code that would show a gb for restricted files in the case where gb-at-request is enabled but file access requests are not allowed.

Example email (from fine logging):

```
Subject: <<<Root: You have requested access to a restricted file>>>. Body: Hello,
You have requested access to a file(s) in dataset "<a href = "http://ec2-3-215-16-135.compute-1.amazonaws.com/dataset.xhtml?persistentId=perma:DANSPIDFK2280271">GBAR</a>". Your request has been sent to the managers of this dataset who will grant or reject your request. If you have any questions, you may reach the dataset managers using the "Contact" link on the upper right corner of the dataset page.

Guestbook Response:


Guestbook Response ID: 283
Response Date: July 25, 2023
Respondent
  Name: Jim Myers
  Email: qqmyers@hotmail.com
  Institution: (No Response)
  Position: reclining
Additional Questions:
  Q: What?
  A: again?



You may contact us for support at noreply@dataverse.yourinstitution.edu.

Thank you,
Root Support|#]
```

mdmADA commented:
I can verify that those updates are all working.

Email to requester and dataset contact:
I am not receiving any emails (even for password reset) to my "_test" account  (the dataset contact) so am unable to check the email that the dataset contact should receive.
I have been able to look at the email sent to the requester.  That looks good.
Can I request that the Dataset title be added to the subject of the email that is sent, at least to the dataset contact. ADA's ticketing systems sorts incoming requests/tickets are sent based on the subject line's dataset title.
( I am not sure how difficult it would be to implement email templates so that dataset owners could specify the details to include in emails sent out by Dataverse).

API:
The API to requestaccess to a datafile whose owner dataset has a guestbook, and guestbook is at request access, is throwing an null pointer exception error.

qqmyers commented:
re: email - the server isn't set up to actually send email afaik. Here's the message to managers:
```
Subject: Root: Access has been requested for a restricted file in dataset "GBAR". 
Body:
 Hello,
File access requested for dataset: GBAR by Jim Myers (qqmyers@hotmail.com). Manage permissions at http://ec2-3-215-16-135.compute-1.amazonaws.com/permissions-manage-files.xhtml?id=141.

You may contact us for support at noreply@dataverse.yourinstitution.edu.

Thank you,
Root Support
```
This example includes the update to have the dataset title in the subject line (same as for the email to the requestor), but is for a gb-at-download setting. Otherwise it would have the same Guestbook text as the previous requestor email.

I also fixed the request access api call (and hopefully therefore the tests that used it).

mdmADA commented:
Regarding the email: I am receiving, eventually as it is quarantined by ANU servers, email to my requester account. So emails ARE being sent.

But no email is sent to my personal account as the dataset contact but perhaps I am not setting my personal account details in the correct  metadata field in order to receive that email. I recall an email from github coming through last week about who receives notifications but I can't find that email now.

The email that I do receive as the requester: the link is coming through as text (ie. all the html markup) rather than as an actual link. Perhaps it is because it is wrapped in quotes? 

The requestaccess api endpoint is no longer throwing an npe.

qqmyers commented:
@mdmADA - I've changed the emails to use html. I don't seem to get emails at all so it is hard to debug formatting issues on this server. Hopefully this change fixes it. If not, I suggest we leave this for when we get this on a QA server. As for recipients - the original access requested email goes to all accounts that can manage the dataset (and hence could grant access). I spoke to loosely saying contacts/owner - there's no connection to the contact metadata and there is no owner of a dataset in Dataverse.

mdmADA commented:
Thanks!

I was testing again (07/08) on your aws instance and it seems that the guestbook at request code is no longer the version deployed?

I am assuming that your code is now waiting for QA testing. Once the code is accepted/merged I will have to do a number of things here to prepare ADA to make the change. If you can let ADA/me know when it is ready, that would be appreciated.

qqmyers commented:
@mdmADA - yes - was working on other branches last week. I've redeployed the GBAR branch for now, but it may disappear again at some point. FWIW: It should be easy to deploy through any of the means for any other branch, e.g. to a local test server. My guess is that this will be part of a 6.1 release (6.0 ~= 5.14 running on Payara6/EE10/Java17/etc.) - no specific ETA yet but that's the thing to watch for/ask about.

sekmiller commented on src/main/webapp/dataverse.xhtml:
```diff
@@ -267,6 +269,21 @@
+                                                  data-toggle="tooltip" data-placement="auto right" data-original-title="#{bundle['dataverse.guestbookentryatrequest.title']}"></span>
+                                        </h:outputLabel>
+                                        <div class="form-col-container">
+                                            <h:selectOneMenu id="dsGuestbookEntryOption" styleClass="form-control" value="#{DataversePage.dataverse.guestbookEntryAtRequest}">
```
This is failing on update because of a property not writable exception. There is a method in the DVObjectContainer for setGuestbookEntryAtRequest but it is updating the guestbookAtRequest field that exists on the Dataverse table

qqmyers replied:
Thanks! FWIW: This was a String versus Boolean issue w.r.t. the type for guestbookEntryAtRequest - the table was originally a String like some other options (e.g. storage driver), but since it is just a flag that can be true/false/null, Boolean was simpler. I missed some places that had to change though. I think it is working now.

sekmiller commented on doc/release-notes/9599-guestbook-at-request.md:
```diff
@@ -0,0 +1,2 @@
+Dataverse can now be configured (via the dataverse.files.guestbook-at-request option) to display any configured guestbook to users when they request restricted file(s) or when they download files (the historic default).
+The global default defined by this setting can be overridden at the collection level on the collection page and at the inidividual dataset level by a superuser using the API. The default - showing guestbooks when files are downloaded - remains as it was in prior Dataverse versions. 
```
individual

sekmiller commented on src/main/java/edu/harvard/iq/dataverse/GuestbookResponse.java:
```diff
@@ -65,8 +67,12 @@ public class GuestbookResponse implements Serializable {
     private AuthenticatedUser authenticatedUser;
 
-    @OneToOne(cascade=CascadeType.ALL,mappedBy="guestbookResponse",fetch = FetchType.LAZY, optional = false)
+    @OneToOne(cascade={CascadeType.REMOVE, CascadeType.MERGE, CascadeType.PERSIST},mappedBy="guestbookResponse",fetch = FetchType.LAZY)
```
It seems like we're associating guestbook responses with file downloads even if they are actually access requests. It will cause confusion on the guestbook response via and download were numbers will be inflated

qqmyers replied:
Rather than trying to change the queries to check for a filedownload entry, I ~reverted #5863 (which was only done in prep for dealing with guestbook-at-request functionality), added 'AccessRequest' as a new eventtype for gb responses and adapted all of the queries to only count non-AccessRequest entries as downloads. (I also changed the names of the queries dealing with downloads to mention download explicitly. There are other queries that still get the total number of gb entries (i.e. ones related to showing the list of all entries) that stayed as they were.) I also adapted the download estimation query which was just using the statistics about table size and average gb entry size to also use the statistic about how many gb entries have eventtype AccessRequest to maintain an accurate count. (In doing this, I raised the threshold at which the estimate starts getting used - before it would kick in with only a few entries which could make download counts be inaccurate when there were only ~tens of downloads which caused some confusion. )

sekmiller commented on src/main/java/edu/harvard/iq/dataverse/GuestbookResponseServiceBean.java:
```diff
@@ -63,30 +63,28 @@ public class GuestbookResponseServiceBean {
                 + " and r.guestbook_id = g.id ";*/
     
-    private static final String BASE_QUERY_STRING_FOR_DOWNLOAD_AS_CSV = "select r.id, g.name, o.id, r.responsetime, f.downloadtype,"
+    private static final String BASE_QUERY_STRING_FOR_DOWNLOAD_AS_CSV = "select r.id, g.name, o.id, r.responsetime, r.downloadtype,"
```
we need to say r.eventtype here

sekmiller commented on src/main/java/edu/harvard/iq/dataverse/GuestbookResponseServiceBean.java:
```diff
@@ -63,30 +63,28 @@ public class GuestbookResponseServiceBean {
     // query has fewer fields than the one above). -- L.A.
-    private static final String BASE_QUERY_STRING_FOR_PAGE_DISPLAY = "select  r.id, v.value, r.responsetime, f.downloadtype,  m.label, r.name "
-                + "from guestbookresponse r, filedownload f, datasetfieldvalue v, filemetadata m , dvobject o "
+    private static final String BASE_QUERY_STRING_FOR_PAGE_DISPLAY = "select  r.id, v.value, r.responsetime, r.downloadtype,  m.label, r.name "
```
also eventtype here
